### PR TITLE
docs: draft for API Best Practices

### PR DESCRIPTION
**Pull request by zcbenz**
This is a rough draft for API Best Practices, nothing here is settled, the purpose of this PR is to collect opinions and get maintainers together to work out a document that most of us agree with.

This document does not mean to be a JavaScript style guide that defines coding styles or naming rules, this means to describe best practices for designing public Electron APIs.

One thing I'm still not sure is to prefer whether getter/setters or properties, as we have discussed in https://github.com/electron/electron/issues/12745 but still have no decision. I have just listed all options here.

Feel free to commit to this branch to fix grammar/wording.

### CODE REVIEW

File: wg-api/best-practices.md [hunk 1]
```diff
@@ -1,4 +1,168 @@
...
+// Constructor
+new BrowserWindow(options)
+
+// Factory method
+nativeImage.createFromPath('/path')
+```
+
+#### Constructor
+
+If the class has inheritance relationship with other public classes, constructor should be used.
```
Comment (codebytere) [inline wg-api/best-practices.md]: nit: `If the class has an inheritance relationship with other public classes, a constructor should be used for that class.`

File: wg-api/best-practices.md [hunk 2]
```diff
@@ -1,4 +1,168 @@
...
+
+```javascript
+if (win instanceof TopLevelWindow) {
+  // ...
+}
+```
+
+#### Factory method
+
+If the class accepts different kinds of parameters with options, we should consider using factory methods.
```
Comment (codebytere) [inline wg-api/best-practices.md]: For consistency: `If the class accepts different kinds of parameters with options, consider using factory methods.`

File: wg-api/best-practices.md [hunk 3]
```diff
@@ -1,4 +1,168 @@
...
+For example:
+
+```javascript
+const { nativeImage } = require('electron')
+nativeImage.createEmpty()
+```
+
+## Class
+
+### Creating Instance
```
Comment (codebytere) [inline wg-api/best-practices.md]: s/`Creating Instance`/`Creating Instances`

File: wg-api/best-practices.md [hunk 4]
```diff
@@ -1,4 +1,168 @@
...
+const { BrowserWindow, nativeImage } = require('electron')
+
+// Constructor
+new BrowserWindow(options)
+
+// Factory method
+nativeImage.createFromPath('/path')
+```
+
+#### Constructor
```
Comment (codebytere) [inline wg-api/best-practices.md]: s/`Constructor`/`Constructors`

File: wg-api/best-practices.md [hunk 5]
```diff
@@ -1,4 +1,168 @@
...
+
+This enables users to detect inheritance relationships between classes.
+
+```javascript
+if (win instanceof TopLevelWindow) {
+  // ...
+}
+```
+
+#### Factory method
```
Comment (codebytere) [inline wg-api/best-practices.md]: s/`Factory method`/`Factory Methods`

File: wg-api/best-practices.md [hunk 6]
```diff
@@ -1,4 +1,168 @@
...
+If a module includes methods or more than one class, the module's name should be in `lowerCamelCase`.
+
+For example:
+
+```javascript
+const { nativeImage } = require('electron')
+nativeImage.createEmpty()
+```
+
+## Class
```
Comment (codebytere) [inline wg-api/best-practices.md]: s/`Classes`/`Class`

File: wg-api/best-practices.md [hunk 7]
```diff
@@ -1,4 +1,168 @@
 # Public API Best Practices
 
-This is a placeholder for a document being drafted by @zcbenz.
-It will be added in a separate PR.
+This is a draft for best practices when designing public APIs.
+
+## Module
```
Comment (codebytere) [inline wg-api/best-practices.md]: s/`Module`/`Modules`

File: wg-api/best-practices.md [hunk 8]
```diff
@@ -1,4 +1,168 @@
...
+
+### Module Name
```
Comment (codebytere) [inline wg-api/best-practices.md]: s/`Module Name`/`Module Names`

File: wg-api/best-practices.md [hunk 9]
```diff
@@ -1,4 +1,168 @@
...
+```javascript
+// Using module methods
+const { nativeImage } = require('electron')
+nativeImage.createFromNamedImage('NSImageNameListViewTemplate', [-1, 0, 1])
+
+// Using static methods
+Buffer.from('7468697320697320612074c3a97374', 'hex')
+```
+
+One reason is ambiguities and usability issues, the `Buffer` API is a good example: [Node.js Buffer API Changes][node-buffer-api].
```
Comment (codebytere) [inline wg-api/best-practices.md]: Let's say:
```
Factory methods help reduce ambuguity and usability issues. The `Buffer` API provides a good example of this: [Node.js Buffer API Changes][node-buffer-api].
````

File: wg-api/best-practices.md [hunk 10]
```diff
@@ -1,4 +1,168 @@
...
+
+Another reason is, due to lack of function overloading, JavaScript class's constructor can only be implemented with a single C++ function, and it is error-prone to parse different kinds of parameters with C++ manually. While factory methods can be perfectly mapped to C++ functions, making the code much easier to read and maintain.
```
Comment (codebytere) [inline wg-api/best-practices.md]: How about: 
> Additionally, due to lack of function overloading, JavaScript class constructors can only be implemented with a single C++ function and are error-prone when parsing different kinds of parameters with C++. Conversely, factory methods can be perfectly mapped to C++ functions, thereby making code much easier to read and maintain.

File: wg-api/best-practices.md [hunk 11]
```diff
@@ -1,4 +1,168 @@
...
+
+### Instance Event
```
Comment (codebytere) [inline wg-api/best-practices.md]: s/`Instance Event`/`Instance Events`

File: wg-api/best-practices.md [hunk 12]
```diff
@@ -1,4 +1,168 @@
...
+
+If the class generates events, it should inherit from `EventEmitter` and use its interface for emitting events.
+
+### Instance Property
```
Comment (codebytere) [inline wg-api/best-practices.md]: s/`Instance Property`/`Instance Properties`

File: wg-api/best-practices.md [hunk 13]
```diff
@@ -1,4 +1,168 @@
...
+
+If an API of the class returns _non-assignable_ `Object`, and the returned objects always strictly equal each other, the API should be implemented as property.
```
Comment (codebytere) [inline wg-api/best-practices.md]: s/`returns _non-assignable_`/`returns a _non-assignable_`

File: wg-api/best-practices.md [hunk 14]
```diff
@@ -1,4 +1,168 @@
...
+For example:
+
+```javascript
+// Should NOT do these, as they have no effect.
+const win = new BrowserWindow(options)
+win.bounds.x = 42
+win.browserViews.push(new BrowserView())
+```
+
+### Instance Method
```
Comment (codebytere) [inline wg-api/best-practices.md]: s/`Instance Method`/`Instance Methods`

File: wg-api/best-practices.md [hunk 15]
```diff
@@ -1,4 +1,168 @@
...
+If an API accepts options, or may accept option in future, it should be implemented as method.
+
+For example:
+
+```javascript
+const win = new BrowserWindow(options)
+win.setBounds(bounds, animate)
+```
+
+### Method vs Property (Option A)
```
Comment (codebytere) [inline wg-api/best-practices.md]: I'll leave edit to this section until after we've discussed them as a WG :)

File: wg-api/best-practices.md [hunk 16]
```diff
@@ -1,4 +1,168 @@
...
+console.log(win.title)
+```
+
+### Method vs Property (Option C)
+
+There is no requirement on whether to use method or property. _(This chapter should be removed)_
+
+## Asynchronous APIs
+
+If an API returns _one_ result asynchronously for each call, it should use `Promise`.
```
Comment (codebytere) [inline wg-api/best-practices.md]: Let's say:
>If an API returns _one_ result asynchronously for each call, it should be implemented with `Promises`.

File: wg-api/best-practices.md [hunk 17]
```diff
@@ -1,4 +1,168 @@
...
+
+For example:
+
+```javascript
+const win = new BrowserWindow(options)
+const result = await win.webContents.executeJavaScript('void 0')
+```
+
+For subscription-style API that returns results for multiple times, it should still use callbacks.
```
Comment (codebytere) [inline wg-api/best-practices.md]: Let's say:
> Subscription-style APIs that returns results multiple times should still be implemented with callbacks.

**Comment by codebytere** [comment]
Thanks for getting this started @zcbenz 🙇‍♀

**Comment by ckerr** [comment]
Maybe we should do something with this PR? :smiley_cat:

**Comment by zcbenz** [comment]
Updated the doc as per last wg-api meeting's discussion.

File: wg-api/best-practices.md [hunk 18]
```diff
@@ -1,4 +1,157 @@
...
+```javascript
+const win = new BrowserWindow(options)
+win.setBounds(bounds, animate)
+```
+
+### Getters and Setters
+
+If the APIs of the class are used to set/get values, they should be implemented as methods.
+
+If the type of the value being set/get is [primitive][primitive], it should be considered also adding an property interface for the value.
```
Comment (jkleinsc) [inline wg-api/best-practices.md]: Considering we want consistency, I think we should be more prescriptive here.  Maybe something like:
```suggestion
If the type of the value being set/get is [primitive][primitive], it should be also added as a property interface for the value.
```

**Pull request merged**